src/service/context/mod.rs

```diff
@@ -151,6 +151,55 @@ impl<S> Context<S> {
     pub fn get<T: Send + Sync + 'static>(&self) -> Option<&T> {
         self.extensions.get::<T>()
     }
+    /// Get a reference to an extension or T's default().
+    ///
+    /// An extension is a type that implements `Send + Sync + 'static`,
+    /// and can be used to inject dynamic data into the [`Context`].
+    ///
+    /// Extensions are specific to this [`Context`]. It will be cloned when the [`Context`] is cloned,
+    /// but extensions inserted using [`Context::insert`] will not be visible the
+    /// original [`Context`], or any other cloned [`Context`].
+    ///
+    /// Please use the statically typed state (see [`Context::state`]) for data that is shared between
+    /// all context clones, parent or not.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use rama::service::Context;
+    /// # let mut ctx = Context::default();
+    /// # ctx.insert(5i32);
+    ///
+    /// // Test existing value
+    /// assert_eq!(ctx.get_or_default::<i32>(), 5i32);
+    ///
+    /// // Test default value for a type not inserted before (assuming `f64::default()` is 0.0)
+    /// assert_eq!(ctx.get_or_default::<f64>(), 0f64);
+    /// ```
+
+    pub fn get_or_default<T: Send + Default + Clone + Sync + 'static>(&self) -> T {
```

Comment:
We do not wish to clone, and we do wish to add. The entire purpose, and sorry if that wasn't clear from the ticket, is that it would also insert `T` using its `Default` implementation if none was defined yet.

So in pseudo code what you want to do is:

1. call just the get method to try to fetch it (this step you do correctly)
2. in case it is found you can return the reference as usual, otherwise continue to step 3
3. call extensions.insert to insert the default value
4. call extensions.get again this time just unwrap as you know it should exist

Notes:

- this will require the method to take `&mut self` as there's a potential mutation
- the previous note will ensure the unwrap is safe
- this also means you no longer need the `Clone` trait bound

[nitpick] I usually put my regular trait bounds in front of the marker trait bounds, so something like `T: Default + Send + Sync + 'static`.

---

Bonus points:

- even better is if you add the same methods first to `Extensions`, as that one wraps around a map, so there you can actually call directly also `get_or`, `get_or_default` and `get_or_else`. This way we never have to call `unwrap` ourselves and do you also simplify your logic :)

Reply:
Can you also add `get_or_else` while you are at it, as that will in generally be preferred over `get_or`, unless you are dealing with a simple primitive value, which will rarely be the case I imagine.